hazardous/metrics/_concordance_index.py

```diff
@@ -0,0 +1,523 @@
+import warnings
+from collections import Counter, defaultdict
+
+import numpy as np
+from scipy.interpolate import interp1d
+
+from .._ipcw import KaplanMeierIPCW
+from ..utils import check_y_survival
+
+
+def concordance_index_incidence(
+    y_test,
+    y_pred,
+    y_train=None,
+    ipcw_estimator="km",
+    time_grid=None,
+    taus=None,
+    event_of_interest=1,
+    tied_tol=1e-8,
+):
+    r"""Time-dependent concordance index for prognostic models with competing risks \
+    using inverse probability of censoring weighting.
+
+    This C-index is event-specific: we set the event of interest as a parameter,
+    and all other event labels are considered to be competing events –except 0, which
+    is the censoring event.
+
+    For example, if ``y_test`` has the following ``"event"`` columns: ``[1, 0, 2, 4]``,
+    and the event of interest is set to ``2``, then we consider the following events:
+    ``[2, 0, 1, 2]``, where 0, 1, 2 respectively denotes the censoring event, the event
+    of interest and competing events.
+
+    We define this C-index as:
+
+    .. math::
+
+        \mathrm{C}(t) = \frac{\sum_{i=1}^n \sum_{j=1}^n (\tilde{A}_{ij}
+        \hat{W}_{ij, 1}^{-1} + \tilde{B}_{ij} \hat{W}_{ij, 2}^{-1}) Q^{ij}(t)
+        \tilde{N}^1_i(t)}
+        {\sum_{i=1}^n \sum_{j=1}^n (\tilde{A}_{ij}
+        \hat{W}_{ij, 1}^{-1} + \tilde{B}_{ij} \hat{W}_{ij, 2}^{-1}) \tilde{N}^1_i(t)}
+
+    where:
+
+    .. math::
+
+        \begin{align}
+        \tilde{N}^1_i(t) &= I\{\tilde{T}_i \leq t, \tilde{D}_i = 1\} \\
+        \tilde{A}_{ij} &= I\{\tilde{T}_i < \tilde{T}_j \cup (\tilde{T}_i =
+        \tilde{T}_j \cap D_j = 0)\} \\
+        \tilde{B}_{ij} &= I\{\tilde{T}_i \geq \tilde{T}_j, D_j = 2\} \\
+        \hat{W}_{ij,1} &= \hat{G}(\tilde{T}_i-|X_i) \hat{G}(\tilde{T}_i|X_j) \\
+        \hat{W}_{ij,2} &= \hat{G}(\tilde{T}_i-|X_i) \hat{G}(\tilde{T}_j-|X_j) \\
+        Q_{ij}(t) &= I\{M(t, X_i) > M(t, X_j)\}
+        \end{align}
```

Comment:
Especially there because you have a mix of \tilde{D}_i and D_j, but both should be the same (observed event, so tilde in Wolbers notation).

Reply:
And we can maybe drop the distinction $\hat{G}(T_i-|X_i)$ vs $\hat{G}(T_i|X_i)$, because I'm not sure we currently take that into account, WDYT ?

Reply:
I agree that we could simplify both expressions